[Site Isolation] Prepare to move ProvisionalFrameProxy ownership from WebFrameProxy to API::Navigation

achristensen07 opened the pull request:
#### dd1da9bd8b6e8790757f9a51e654563481e151b1
<pre>
[Site Isolation] Prepare to move ProvisionalFrameProxy ownership from WebFrameProxy to API::Navigation
<a href="https://bugs.webkit.org/show_bug.cgi?id=274604">https://bugs.webkit.org/show_bug.cgi?id=274604</a>
<a href="https://rdar.apple.com/128626098">rdar://128626098</a>

Reviewed by Charlie Wolfe.

In order to fix the FIXME comment at the bottom of the SiteIsolation.CancelProvisionalLoad API test,
we need to be able to be cancelling a provisional load in a process at the same time as we are starting
a new provisional load in that same process.  In order to make that work, the ProvisionalFrameProxy
can&apos;t be destroyed when the first provisional load fails.  We will need to move the ProvisionalFrameProxy
ownership to the Navigation, and introduce the possibility of multiple Navigation objects keeping it
alive, so it needs to be RefCounted instead of owned by a unique_ptr.  We also need the message
WebPage::DestroyProvisionalFrame to be sent from the UI process instead of called directly by the web
process because at the time a provisional load failure is being dispatched from the web process,
the web process doesn&apos;t know whether the provisional frame should be destroyed.  I set
loadParameters.navigationID in a few more places so we can get the Navigation object in more places.
I remove the use of WebFrameProxy::provisionalFrame in WebPageProxy::decidePolicyForResponse in favor
of getting the WebProcessProxy from the IPC::Connection, which requires adding a way to get a
WebProcessProxy from an AuxiliaryProcessProxy.  With site isolation on, we will need to set up navigation
identifiers for all frame loads instead of just the main frame, so I do that behind a siteIsolationEnabled
check.  WebFrameProxy::prepareForProvisionalLoadInProcess will need to modify the navigation by setting
its ProvisionalFrameProxy, so I make it non-const.

This is all the set-up I can do without really changing behavior.  An upcoming PR will do the rest of the
more invasive changes to get that test addition to work completely.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::setProvisionalFrame):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::provisionalFrame):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::create):
(WebKit::ProvisionalFrameProxy::~ProvisionalFrameProxy):
(WebKit::ProvisionalFrameProxy::takeFrameProcess):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::takeProvisionalFrame):
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::decidePolicyForResponse):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.h:
(isType):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::createDocumentLoader):

Canonical link: <a href="https://commits.webkit.org/279291@main">https://commits.webkit.org/279291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b49a22cc55e4d2fee41fa2c5ffc392aceaa735d6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52846 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32183 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5333 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56125 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3569 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55151 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39115 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3409 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42898 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2307 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54944 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30043 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45617 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24004 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27091 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2934 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1728 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48947 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57718 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50281 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29207 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45809 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49644 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30126 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7794 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28961 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
```diff
@@ -41,11 +47,16 @@ ProvisionalFrameProxy::ProvisionalFrameProxy(WebFrameProxy& frame, Ref<FrameProc
+ProvisionalFrameProxy::~ProvisionalFrameProxy()
+{
+    if (m_frameProcess && m_frame->page())
+        m_frame->page()->sendToWebPageInProcess(m_frameProcess->process(), Messages::WebPage::DestroyProvisionalFrame(m_frame->frameID()));
```
I don't understand why we need this separate function. Can't we just use `m_frameProcess->process().send()`?

achristensen07 replied:
We could, but we would need a way to get the WebCore::PageIdentifier to send to.  That logic is already in WebPageProxy:: sendToWebPageInProcess

achristensen07 committed:
[Site Isolation] Prepare to move ProvisionalFrameProxy ownership from WebFrameProxy to API::Navigation
https://bugs.webkit.org/show_bug.cgi?id=274604
rdar://128626098

Reviewed by Charlie Wolfe.

In order to fix the FIXME comment at the bottom of the SiteIsolation.CancelProvisionalLoad API test,
we need to be able to be cancelling a provisional load in a process at the same time as we are starting
a new provisional load in that same process.  In order to make that work, the ProvisionalFrameProxy
can't be destroyed when the first provisional load fails.  We will need to move the ProvisionalFrameProxy
ownership to the Navigation, and introduce the possibility of multiple Navigation objects keeping it
alive, so it needs to be RefCounted instead of owned by a unique_ptr.  We also need the message
WebPage::DestroyProvisionalFrame to be sent from the UI process instead of called directly by the web
process because at the time a provisional load failure is being dispatched from the web process,
the web process doesn't know whether the provisional frame should be destroyed.  I set
loadParameters.navigationID in a few more places so we can get the Navigation object in more places.
I remove the use of WebFrameProxy::provisionalFrame in WebPageProxy::decidePolicyForResponse in favor
of getting the WebProcessProxy from the IPC::Connection, which requires adding a way to get a
WebProcessProxy from an AuxiliaryProcessProxy.  With site isolation on, we will need to set up navigation
identifiers for all frame loads instead of just the main frame, so I do that behind a siteIsolationEnabled
check.  WebFrameProxy::prepareForProvisionalLoadInProcess will need to modify the navigation by setting
its ProvisionalFrameProxy, so I make it non-const.

This is all the set-up I can do without really changing behavior.  An upcoming PR will do the rest of the
more invasive changes to get that test addition to work completely.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebKit/UIProcess/API/APINavigation.cpp:
(API::Navigation::setProvisionalFrame):
* Source/WebKit/UIProcess/API/APINavigation.h:
(API::Navigation::provisionalFrame):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::create):
(WebKit::ProvisionalFrameProxy::~ProvisionalFrameProxy):
(WebKit::ProvisionalFrameProxy::takeFrameProcess):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::takeProvisionalFrame):
(WebKit::WebFrameProxy::prepareForProvisionalLoadInProcess):
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::decidePolicyForResponse):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.h:
(isType):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::createDocumentLoader):

Canonical link: https://commits.webkit.org/279291@main
